### ISSUE TITLE
Question: How do determine an update is available?

### ISSUE THREAD

**Issue by JAStanton** [Microsoft/react-native-code-push#54]
This seems silly but I just dont understand it:

I have deployed a version with version number: `1.0.1` (Email me for the sha's)

If I hit your api like this:
```
https://codepush.azurewebsites.net/updateCheck?deploymentKey=RECACTED&appVersion=0.1.0&packageHash=&isCompanion=
```

I get this response:
```json
{
  "updateInfo": {
    "downloadURL": "",
    "description": "",
    "isAvailable": false,
    "isMandatory": false,
    "appVersion": "1.0.1",
    "packageHash": "",
    "label": "",
    "packageSize": 0,
    "updateAppVersion": true
  }
}
```

if I hit the api like this: 
```
https://codepush.azurewebsites.net/updateCheck?deploymentKey=RECACTED&appVersion=1.0.1&packageHash=&isCompanion=
```

```json
{
  "updateInfo": {
    "downloadURL": "https://codepush.blob.core.windows.net/storagev2/REDACTED",
    "description": "Another test",
    "isAvailable": true,
    "isMandatory": false,
    "appVersion": "1.0.1",
    "packageHash": "REDACTED",
    "label": "v2",
    "updateAppVersion": false
  }
}
```

So to summarize:
I need to update if my client is `1.0.1` and my latest build is `1.0.1`, this seems odd.
also if I have an older version `0.1.0` I do need to update but I don't have any download urls and the code actually bails on updating on this [line of code](https://github.com/Microsoft/react-native-code-push/blob/ff891dba769f979a453759519ce9af0a5788d46f/CodePush.ios.js#L100-L104)

What's going on here?

**Comment by JAStanton** [comment]
Piggy backing off of this, different scenario:

1) I pushed an update with build of `1.0.0`
2) I update my plist with `0.1.0` 
3) I build and launch my app and I call `checkForUpdates` and it comes back as null.

Following the logic I see my response from the server is this: 
(nothing has been redacted, this is the exact response)
```json
{
  "updateInfo": {
    "downloadURL": "",
    "description": "",
    "isAvailable": false,
    "isMandatory": false,
    "appVersion": "1.0.1",
    "packageHash": "",
    "label": "",
    "packageSize": 0,
    "updateAppVersion": true
  }
}
```

It think there is an update but hits this line
`Ignore updates that require a newer app version, since the end-user couldn't reliably apply it`

and aborts. 
Thoughts?

**Comment by lostintangent** [comment]
When querying the service for an update, it can respond in one of three ways:

1. There is an update and here is the download URL for it. This only happens when there the latest release for the deployment has a different hash then the client, but the exact same binary version (which is what we mean by "App Store version")

2. There is an update, but it is associated with a newer version of the binary than you are running. This happens when the latest release for a deployment has a different hash then the client is running, but the release is tagged with a binary version that is greater than what the client is running. In this case, the server doesn't provide a download URL under the assumption that the client needs to update the binary before they could reliably apply the script updates.

As you illustrated, the service does return that an update is available in this case, but both clients currently translate this scenario into a "no update is available". At some point we need to add support here so that devs could choose to display a "Go to the store to update" message

3. The latest release for the deployment has a different hash then what the client is running, but it's tagged version is older that what the client is running. In this case, the server responds saying that there isn't an update available, based on the assumption that the user is running a binary that actually has newer bits than CodePush has

The scenario you mentioned that you want to support should therefore work just fine. That said, we had a temporary bug with this feature that we fixed this afternoon so it might be worth trying again to see if it's resolved :)

**Comment by lostintangent** [comment]
Did that answer your question at all? Let me know how we can help. Thanks!

**Comment by rhaker** [comment]
I've been using CodePush for the past few days (thanks for putting this together!) and will provide more feedback under another issue. But I'm trying to understand the updating behavior and I think I have the same confusion as discussed above.

To be clear, if I build a version 0.1.7 in Xcode which gets uploaded to the App Store, then the plain vanilla case for applying patches (via CodePush) would be to apply all code-push release updates as 0.1.7? In other words, if I make a tiny text change and release it to CodePush as version 0.1.8, then the update won't be applied to users who installed 0.1.7 via the App Store because there's an assumption that the App Store Binary may be incompatible and needs to be updated? 

Is my understanding of this correct? Thanks in advance for all the help (and I'll provide more feedback under another issue)...

**Comment by lostintangent** [comment]
@rhaker Your assumption is correct. As of right now, we expect that all CodePush deployed updates are targeted at a specific app store/binary version. As you mentioned in #80, I'm going to close this issue. 

If you have a need for a change in this behavior, please let us know and open up another issue that is specific to your scenario. Thanks!

**Comment by nevir** [comment]
> There is an update, but it is associated with a newer version of the binary than you are running. This happens when the latest release for a deployment has a different hash then the client is running, but the release is tagged with a binary version that is greater than what the client is running. In this case, the server doesn't provide a download URL under the assumption that the client needs to update the binary before they could reliably apply the script updates.

Hmm, @lostintangent, this seems like dangerous behavior to me. For example, let's say we have the following builds:

* binary: 1.0, bundle: 1
* binary: 1.0, bundle: 2
* binary: 1.1, bundle: 3

Say a user has `binary: 1.0, bundle: 1`.  As described, when that app checks for updates, the server will respond with "no updates".  However, I think it would be preferable for the server to return with a pointer to bundle 2 (the most recent bundle available for that binary version).

Going a step further, the server could also return an extra bit of metadata with that response indicating that there's also a newer binary version

**Comment by lostintangent** [comment]
I 100% agree with you. We've been discussing the concept of a deployment being "forked" every time you release an update with a new target App Store version. That way, it would be easier to understand/visualize the release history (as well as things like rollback) within a deployment that is potentially maintaining multiple binary versions in parallel. 

Currently, our recommendation is to create a new deployment when you release a new App Store version, if you want to achieve the behavior you recommended. It's a little bit of a hassle to do, but it's at least easy to understand. When you release a new App Store version, you rename the current deployment to "Production-{Version}", create a new production one and add the new deployment key to the app before submitting it to the store. That way, you can release to each one, knowing that your users will absolutely get the updates.

That solution is effectively the "manual" way of achieving a deployment forking solution we could offer by default. We have many users doing this currently and it works well, but yes, is a pain. I'm taking your feedback as a catalyst to rediscuss this again since I agree we can do much better here.

Regarding your question about the client being notified about a newer App Store version, the server actually already does this, but the client currently ignores it. We plan to change this such that sync will fire a new status event indicating the precense of a newer App Store version and the app can choose to do what they want with that information.

We're trying to move as quick as we can, but we obviously have some more work to do :) thank you so much for the feedback! Let me know if any of this behavior will be a blocker for you guys and we will prioritize accordingly.

**Comment by nevir** [comment]
Awesome, yeah. I think deployment forking makes a lot of sense.

Also: an alternative to think about is expressing range constraints on the binary version rather than exact versions (but I don't think I understand all the implications of that yet)

None of this is currently blocking us, though, since we've got a separate version check system in place already on our API side that can help us handle users who end up with extremely stale builds

**Comment by rhaker** [comment]
Jonathan, are people really doing that approach? It makes me think of just creating a CodePush app for every binary and then the issue above doesn't occur.

Also, a few thoughts. Whenever you submit a new binary, both App Stores should notify the user that an update is available. Now, of course, users ignore this - which is why the CodePush approach is better (with a forced update), but I'm not sure how dangerous it is. Having a message popup telling the user of a new update is probably enough.

**Comment by lostintangent** [comment]
Yep you could choose to create a new app instead of a new deployment. As you mentioned, either case is a bit of a pain :)

We want to notify the client about a new App Store version, and allow the app to do with that information what they will. We're not sure what the right experience/policy is in that scenario, but we at least think it makes sense to provide apps with the data.

The version range definitely makes a lot of sense as well. I'm going to give that some thought this evening, to see whether that isn't a simple solution for this scenario. Thanks for mentioning that!

**Comment by lostintangent** [comment]
Hmm, some of my comments above were responding to @rhaker's comment that seems to have been deleted :)
